### PR TITLE
Document, test and improve unssuported scenarios

### PR DESCRIPTION
**Pull request by jreidinger**
related trello: https://trello.com/c/Wj8zx82g/1781-8-yast-sudo-new-better-parser

Goal of this PBI is to document what is supported and what not. Add example sudoers file that can be used to testing its ability. Also adding unit test for parsing part. And improve three things:

1. write properly `@include` and `@includedir`
2. improve error report when write failed
3. stop when unsupported scenario is detected and inform user

### Screenshots

old error report when write failed:
sudo_fail_old
new one with better error message:
sudo_fail_new
Unsupported multiple tags:
sudo_unsupported
Unsupported digest:
sudo_unsupported2

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+## YaST - Sudo Configuration
+
+Intention of this module is to provide UI for configuring sudo.
+
+### Known Limitations
+
+- Module using for parsing sudoers own handcrafted perl parser which has
```
Comment (kobliha) [inline README.md:7]: Something like: We use own hand-crafted parser written in Perl for reading and writing the configuration. It has some limitations especially regarding changes in configuration format introduced in sudo X.Y.Z. Augeas lenses are currently more up-to-date.

File: test/data/cmd_alias_config/sudoers [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+			/usr/sbin/rrestore, /usr/bin/mt
+Cmnd_Alias	KILL = /usr/bin/kill
+Cmnd_Alias	PRINTING = /usr/sbin/lpc, /usr/bin/lprm
+Cmnd_Alias	SHUTDOWN = /usr/sbin/shutdown
+Cmnd_Alias	HALT = /usr/sbin/halt
+Cmnd_Alias	REBOOT = /usr/sbin/reboot
+Cmnd_Alias	SHELLS = /sbin/sh, /usr/bin/sh, /usr/bin/csh, /usr/bin/ksh, \
+			 /usr/local/bin/tcsh, /usr/bin/rsh, \
+			 /usr/local/bin/zsh
+Cmn_Alias	SU = /usr/bin/su
```
Comment (kobliha) [inline test/data/cmd_alias_config/sudoers:21]: I guess `Cmnd_Alias` is what you wanted

File: test/data/cmd_alias_config/sudoers [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+Cmn_Alias	VIPW = /usr/sbin/vipw, /usr/bin/passwd, /usr/bin/chsh, \
```
Comment (kobliha) [inline test/data/cmd_alias_config/sudoers:22]: And here too

File: README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  limitations especially regarding new sudo changes as opposite to e.g.
+  augeas lenses that are more up-to data.
+- Support for `@include` and `@includedir` directive is limited. It just
+  shows them in rules section, but nothing more. Also duplicite alias detection
+  does not work across include files.
+- Sudo configuration does not support multitag in rule, which leads to dropping
+  it ( caused by limitation of parser, see above )
```
Comment (kobliha) [inline README.md]: Sounds like a bug. Can we prevent from that?

**Review by kobliha — CHANGES_REQUESTED** [review]

File: test/data/cmd_alias_config/sudoers (hunk 2 above)
Comment (jreidinger) [reply test/data/cmd_alias_config/sudoers:21]: ah, I want here in fact `Cmd_Alias` to demonstrate that command aliasing

File: README.md (hunk 4 above)
Comment (jreidinger) [reply README.md]: OK, so parser is not limitation here as it return whole part as string https://github.com/yast/yast-sudo/blob/master/src/servers_non_y2/ag_etc_sudoers#L30
problem is in fact in separt part of parsing done in module which read just one TAG:
https://github.com/yast/yast-sudo/blob/master/src/modules/Sudo.rb#L175
and of course list of tags is limited, sudo supports much more tags.

Comment (jreidinger) [reply README.md]: I try to fix it, but failed misserably, as I found another limitation with command specific tags and for it, we do not have prepared data structures ( we have tags list and command list, but not that tag is associated with command ).

**Review by kobliha — APPROVED** [review]
LGTM

File: README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+The intention of this module is to provide a User Interface for configuring
+`sudo`.
+
+### Known Limitations
+
+- It uses a handcrafter Perl parser that has some limitations, especially when
+  it comes to deal with new `sudo` features or complex configuration.
+  Alternatives like Augeas lenses are more up-to-date. For specific limitations
+  see below
```
Comment (teclator) [inline README.md:11]:
```suggestion
  see below.
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+- Support for `@include` and `@includedir` directive is limited. It just shows
+  them in rules section, but nothing more. Also duplicated alias detection does
+  not work across included files.
+- `sudo` configuration does not support multitags in rules, which leads to
+  refuse to work with error message
```
Comment (teclator) [inline README.md:16]:
```suggestion
  refuse to work with error message.
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+- `sudo` configuration does not support command specific tags in rules, which leads to
+  refuse to work with error message
```
Comment (teclator) [inline README.md:18]:
```suggestion
  refuse to work with error message.
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+- The module does not allow to see/edit the global configuration of `sudo` (key
+  `Defaults`).
+- No support for commands digest feature. If found then it refuse to work.
```
Comment (teclator) [inline README.md:21]:
```suggestion
- No support for commands digest feature. If found then it refuses to work.
```

File: package/yast2-sudo.changes [hunk 9]
```diff
@@ -1,3 +1,13 @@
+-------------------------------------------------------------------
+Thu Oct  8 14:43:17 UTC 2020 - Josef Reidinger <jreidinger@suse.com>
+
+- Support @include(-dir) directives
+- Support alternative name Cmd_Alias
+- report properly if some configuration yast2-sudo cannot read
```
Comment (teclator) [inline package/yast2-sudo.changes:6]:
```suggestion
- report properly if yast2-sudo cannot read some configuration
```

**Comment by teclator** [comment]
In general LGTM, and it was already approved by @kobliha , so, feel free to apply the suggestions and/or merge it

File: src/modules/Sudo.rb [hunk 10]
```diff
@@ -566,7 +589,16 @@ def PollAbort
     def Read
       return false if !Confirm.MustBeRoot
 
-      Report.Error(Message.CannotReadCurrentSettings) if !ReadSudoSettings2()
+      begin
+        Report.Error(Message.CannotReadCurrentSettings) if !ReadSudoSettings2()
+      rescue UnsupportedSudoConfig => e
+        msg = _("Unsupported configuration found. YaST2 exits now to prevent breaking system.")
```
Comment (kobliha) [inline src/modules/Sudo.rb:595]: What about `YaST will not exit to prevent from breaking the system.`?

**Comment by kobliha** [comment]
One little comment and LGTM :+1:

File: src/modules/Sudo.rb (hunk 10 above)
Comment (jreidinger) [reply src/modules/Sudo.rb:595]: fine for me

**Pull request merged**